woocommerce/woocommerce-gutenberg-products-block
# Remove `useStoreNotices` and interact directly with data store instead

This PR will remove a lot of the items within `StoreNoticesContext` (except one value - `isSuppressed` - which will be moved later when #6155 is closed). The notice system is provided by `@wordpress/notices` and a data store from `@wordpress/data` so we can interact directly with this store instead of confusing hooks and contexts.

- Remove `useStoreNotices` hook.
- Strip out everything in `StoreNoticesContext` besides `isSuppressed`.
- Change value in the context to only `isSuppressed` and `setIsSuppressed`.
- Add a util function (`hasNoticesOfType`) that will check if a notice for a given context and given type exists in the `core/notices` store.
- Add a util function (`removeNoticesByStatus`) which will remove all notices with a given status.
- Add tests for `hasNoticesOfType` and `removeNoticesByStatus`.
- Make `StoreNoticesContainer` a named export, so it can be imported from `@woocommerce/base-context`.
- Update all uses of `useStoreNotices` so that it gets data from the `core/notices` store directly.
- Add a `StoreNoticeContainer` to the All Products block and the Single Product block.
- Set contexts in the error notices generated by both of these blocks.


Fixes #6100

### Testing

#### Automated Tests
### Manual Testing

How to test the changes in this Pull Request:

#### Testing scenario 1
1. Add a coupon to your store.
2. Add items to your cart.
3. Go to the Cart block, and apply the coupon. Ensure you see the snackbar notice for successful coupon application.
4. **In a new tab**, open the Cart block again and remove the coupon. Ensure you see the snackbar notice for successful coupon removal.
5. Go back to the original tab and try removing the coupon, ensure you see an error message above the Cart.
6. Repeat the same steps but for the Checkout block.

#### Testing scenario 2
1. Go to the Checkout block and enter an invalid address, for example make the postcode an incorrect format for the country you've selected.
2. Ensure an error appears above the block.
3. Correct the error and ensure the notice disappears.

### Testing scenario 3

1. Enable Express Payment Methods, this is easiest done using the stripe plugin.
2. Add items to your cart and go to the Checkout block.
3. Enter an invalid postcode (for example in the UK enter `123`).
4. When the error appears above the block, click the express payment button. The error should be hidden while the express payment method modal is open.
5. Close the modal and expect the errors to reappear.

### Testing scenario 4
1. Make **two items** in your store only be able to be sold individually by checking the `Enable this to only allow one of this item to be bought in a single order` checkbox in the inventory section of the Edit product page.
2. Go to the All Products block.
3. Add this item to your cart.
4. Expect no errors.
5. Add the same item.
6. Expect an error at the top of the page.
7. Add it again, while the request is processing the error should disappear.
8. Create a new page and add a Single Product block for _each_ product you set to be individually sold above.
9. Visit the page and try to add each product more than once. Ensure an error appears for each block, and they are independent of each other (can be cleared independently etc.)
10. Try to add one of the products to the cart while the error is showing, it should disappear during the request.

#### Testing scenario 5
1. This requires you to have a development copy of the Stripe plugin, go here and make it return an error message instead of an empty string.
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8f181abe9ad8321ebdd09a494ed0e570b2f2c448/client/blocks/payment-request/hooks.js#L111
2. Go to the Cart block and ensure you have express payments set up.
11. Click the express payment button, ensure no errors appear while the modal is displaying.
12. Close the modal without paying and ensure you see your error message in the express payment method section.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
* [x] Same as above, **EXCEPT SCENARIO 5**
